### ISSUE TITLE
Allow delegation as part of google-auth library

### ISSUE THREAD

**Issue by salrashid123** [googleapis/google-cloud-common#266]
Feature request to add in delegation support into google-auth libraries.

If i enable delegation on a service account, i can run through `iamcredentials.generateAccessToken()` and then use that raw_token in an api call or credentials.

This FR is to make that step easier within google-auth library set.

for example, right now i get a credential like this:
```
    credentials, project = google.auth.default(scopes=scopes)    
```

which will retrun the default credential

however, what i really want to get is a token the default credential has access through via impersonation.

in the example below, i added in the `act_as=` parameter to signal the credential i really want.  The resulting credential object is for the destination account

```
credentials, project = google.auth.default(
      scopes = ['https://www.googleapis.com/auth/devstorage.read_only'], 
      act_as='impersonated-account@fabled-ray-104117.iam.gserviceaccount.com')
```

i'm attaching a hack to the google-auth library that i've verified works in service_account credentials and on GCE

to use
```
# virtualenv env
# source env/bin/activate
# pip install google-cloud-storage==1.13.0 google-auth==1.5.1
# cp _default.py env/lib/python2.7/site-packages/google/auth/_default.py
```

-   main.py  https://gist.github.com/salrashid123/537bbac0e4dcc03eeb8efeadad32072d
- _default.py https://gist.github.com/salrashid123/5674def6a8233661bc7db1629f87d5d4

on local
export GOOGLE_APPLICATION_CREDENTIALS=/path/to/svc_account.json

---

on gce, create instance w/ just IAM

```
  gcloud beta compute  instances create instance-1 \
    --zone=us-central1-a  \
    --service-account=1071284184436-compute@developer.gserviceaccount.com \
    --scopes=https://www.googleapis.com/auth/iam
```

then ssh in and the same steps w/ virtualenv,etc

(sample does not do anything w/ delegates=[], etc; i also 'just added' the access_token into the credential (i don't know if there is an [oauth2client.client.AccessTokenCredential](https://oauth2client.readthedocs.io/en/latest/source/oauth2client.client.html) available

**Comment by JustinBeckwith** [comment]
Making sure @theacodes has a chance to chime in on this one :)

**Comment by theacodes** [comment]
We already have this.

https://google-auth.readthedocs.io/en/latest/reference/google.oauth2.service_account.html#domain-wide-delegation

**Comment by salrashid123** [comment]
@theacodes 
No..domain-wide delegation is entirely different thing.  This is about one service account acting as another unrelated to that gsuites artifact.

**Comment by theacodes** [comment]
I see. This would be better implemented in the same way as domain-wide delegation, as in adding `act_as` and `with_act_as` to `google.oauth2.service_account.Credentials`.

**Comment by salrashid123** [comment]
Got it, i'll see if i can setup a PR (may take sometime to get the unit tests going)

note, this capability also works for a 'user' credential acquired via app defaults.

The issue with using user accounts is the ADC default project 
```
$ cat google-cloud-sdk/lib/googlecloudsdk/api_lib/auth/util.py
DEFAULT_CREDENTIALS_DEFAULT_CLIENT_ID = '764086051850-6qr4p6gpi6hn506pt8ejuq83di341hur.apps.googleusercontent.com'
```

doesnt' have the `iamcredentials` api enabled.   That is, if you run the sample above w/ a non-service account (i.,e as you), the return message is:

```json
{    "message": "Your application has authenticated using end user credentials from the Google Cloud SDK or Google Cloud Shell which are not supported by the iamcredentials.googleapis.com. We recommend that most server applications use service accounts instead. For more information about service accounts and how to use them in your application, see https://cloud.google.com/docs/authentication/.",
    "status": "PERMISSION_DENIED"
  }
```

you can set `ADC` to use a differnet `client_id/client_secret` from a project that does have iamcredentials enabled 
```
gcloud auth application-default login --client-id-file=/path/to/client_secret_installed.json
```

after that even a user can impersonate the service account throuh `google-auth`  
I'll mention this in the docs but I  do recognize  the primary usecase for this FR with service accounts

**Comment by theacodes** [comment]
I consider that a non-problem as we're expecting to remove gcloud as an ADC
provider eventually.
